### PR TITLE
Fixes response when QAT returns QZ_BUF_ERROR

### PR DESCRIPTION
**Pull request by parkerha1**
Changes the Reader and the Writer to still use the processed results even after a buffer error from QAT.
Changes the normal resizing to be done after the output buffer draining, with the original resizing still being done if we get a buffer error with no processed results.
Adjusts the use of z.bufferGrowth to align with the BufferGrowthOption documentation. Before, it was being as a flat add, now it is used as a scaling of the previous size.

Upgrades qgzip dependency on qatgo

### CODE REVIEW

File: qatzip/options.go [hunk 1]
```diff
@@ -238,19 +238,18 @@ func BounceBufferLengthOption(len int) Option {
 	}
 }
 
-// If output buffer is too small (see QZ_BUF_ERROR) increase size of output buffer a factor of len and retry
+// If output buffer is too small (see QZ_BUF_ERROR) increase size of output buffer a factor of mult and retry
```
Comment (markbrown314) [inline qatzip/options.go]: We should not do an API change here. Could potentially change the expected behavior of existing code.

File: qatzip/params.go [hunk 2]
```diff
@@ -12,8 +12,9 @@ type InputBufferMode int
 const (
 	DefaultCompression        = 1
 	MinBufferLength           = 128 * 1024
+	MaxBufferLength           = 2 * 1024 * 1024 * 1024
 	DefaultBufferLength       = 128 * 1024 * 1024
-	DefaultBufferGrowth       = 1024 * 1024
+	DefaultBufferGrowth       = 2
```
Comment (markbrown314) [inline qatzip/params.go]: Instead of multiplying we can just increment in 128MB steps.

File: qatzip/decompress.go [hunk 3]
```diff
@@ -137,7 +138,20 @@ func (z *Reader) Read(p []byte) (n int, err error) {
 			z.outputBufLeft -= np
 			produced += np
 			remainder -= np
-			continue
+			if !isPartialJob || produced == len(p) {
```
Comment (markbrown314) [inline qatzip/decompress.go]: You do not need to expand the output buffer in the drain buffer code.

File: qatzip/decompress.go [hunk 4]
```diff
@@ -150,8 +164,8 @@ func (z *Reader) Read(p []byte) (n int, err error) {
 				return produced, io.EOF
 			}
 
-			rr := trace.StartRegion(z.ctx, "Qz(3) Input Stream")
 			for !z.streamDone {
+				rr := trace.StartRegion(z.ctx, "Qz(3) Input Stream")
```
Comment (markbrown314) [inline qatzip/decompress.go]: If you move the trace to the inner loop. It will produce a separate trace for every Read call. That was not the intention.

File: qatzip/decompress.go [hunk 5]
```diff
@@ -195,19 +210,24 @@ func (z *Reader) Read(p []byte) (n int, err error) {
...
-				// TODO grow to a maximum size
+			if err == ErrBuffer && !isPartialJob {
+				// Grow output buffer, only in the case that it was not large enough for QAT to output any data.
+				// If QAT does output data, we save a copy by doing it after the output buffer draining.
+				
 				t1 = time.Now().UnixNano()
-				z.bufferGrowth *= 2
-				z.traceLogf(Med, "[expand output buffer] obl:%v -> %v", len(z.outputBuf), remainder+z.bufferGrowth)
-				z.outputBuf = make([]byte, remainder+z.bufferGrowth)
+				newLength := len(z.outputBuf)*z.bufferGrowth
```
Comment (markbrown314) [inline qatzip/decompress.go]: Change this from a multiply to an add. See below about API change.

File: qatzip/compress.go [hunk 6]
```diff
@@ -340,20 +342,24 @@ func (z *Writer) compressWrite(p []byte) (n int, err error) {
...
+			if err == ErrBuffer && !isPartialJob {
+				// Grow output buffer, only in the case that it was not large enough for QAT to output any data.
+				// If QAT does output data, we save a copy by doing it after the output buffer draining.
+
 				t1 = time.Now().UnixNano()
-				z.bufferGrowth *= 2
-				newSize := remainder + z.bufferGrowth
-				z.traceLogf(Med, "[expand output buffer] o:%v n:%v", z.outputBuf.Len(), newSize)
-				z.outputBuf = bytes.NewBuffer(make([]byte, newSize))
+				newLength := z.outputBuf.Len()*z.bufferGrowth
```
Comment (markbrown314) [inline qatzip/compress.go]: I mentioned below that we should not change the API. I think we can get away with just adding buffer growth value to the size of output buffer. There is an edge case we need to test for when the input size is 2GB and it has no way to expand it could cause a infinite loop. You need to just return an error in that case.

File: qatzip/decompress.go [hunk 7]
```diff
@@ -195,19 +210,24 @@ func (z *Reader) Read(p []byte) (n int, err error) {
...
+				if newLength > MaxBufferLength {
+					newLength = MaxBufferLength
+				}
```
Comment (markbrown314) [inline qatzip/decompress.go:207]: There is an edge case we need to test for when the input size is 2GB and it has no way to expand it could cause a infinite loop. You need to just return an error in that case.

File: qatzip/params.go [hunk 8]
```diff
@@ -12,8 +12,9 @@ type InputBufferMode int
 const (
 	DefaultCompression        = 1
 	MinBufferLength           = 128 * 1024
+	MaxBufferLength           = 2 * 1024 * 1024 * 1024
```
Comment (markbrown314) [inline qatzip/params.go]: I think we should make this configurable for the case where Go increases the slice limitation and if the user is memory constrained and wants a smaller maximum. Just check that Min <= Max.

File: qatzip/compress.go [hunk 9]
```diff
@@ -340,20 +342,24 @@ func (z *Writer) compressWrite(p []byte) (n int, err error) {
 		z.traceLogf(Med, "[write->qat] r:%v i:%v o:%v ibofs:%v obl:%v err:%v", remainder, in, out, consumed, z.outputBuf.Len(), err)
 
 		if err != nil {
-			if err == ErrBuffer {
-				// expand output buffer
-				// TODO grow to a maximum size
+			if err == ErrBuffer && !isPartialJob {
```
Comment (markbrown314) [inline qatzip/compress.go]: I don't think we need to change this. I have only seen ErrBuffer in the case where the Output Buffer is too small in compression.

File: qatzip/decompress.go [hunk 10]
```diff
@@ -175,15 +189,16 @@ func (z *Reader) Read(p []byte) (n int, err error) {
...
-			rr.End()
 		}
 
 		// decompress input data
 		rq := trace.StartRegion(z.ctx, "Qz(2) Decompress")
 		t1 = time.Now().UnixNano()
 		in, out, err := z.q.Decompress(z.inputBuf[z.inputBufOffset:z.inputBufRead], z.outputBuf)
-		if err == nil {
+		isPartialJob = (out > 0 && err == ErrBuffer)
+		if err == nil || isPartialJob {
```
Comment (markbrown314) [inline qatzip/decompress.go]: You can replace isPartialJob with err == ErrBuffer

File: qatzip/decompress.go [hunk 11]
```diff
@@ -195,19 +210,24 @@ func (z *Reader) Read(p []byte) (n int, err error) {
 			in, out, len(z.inputBuf), z.inputBufOffset, z.inputBufRead, len(z.outputBuf), err)
 
 		if err != nil {
-			if err == ErrBuffer {
-				// expand output buffer
-				// TODO grow to a maximum size
+			if err == ErrBuffer && !isPartialJob {
```
Comment (markbrown314) [inline qatzip/decompress.go]: Replace isPartialJob with out == 0

**Review by markbrown314 — CHANGES_REQUESTED** [review]
Make sure you tag the issue # in the commit message.

File: qatzip/qatzip_internal.c [hunk 12]
```diff
@@ -418,7 +418,7 @@ int qatzip_decompress(qatzip_state_t * state, unsigned char *in_buf, unsigned in
 		     stream->pending_in, stream->pending_out, status);
 	qatzip_debug_dump(QDL_DEBUG, state, stream->out, stream->out_sz);
 
-	if (status != QZ_OK) {
+	if (status != QZ_OK || status != QZ_BUF_ERROR) {
```
Comment (markbrown314) [inline qatzip/qatzip_internal.c]: This should be &&

File: qatzip/decompress.go [hunk 13]
```diff
@@ -195,19 +195,25 @@ func (z *Reader) Read(p []byte) (n int, err error) {
...
 
 		if err != nil {
-			if err == ErrBuffer {
-				// expand output buffer
-				// TODO grow to a maximum size
+			if err == ErrBuffer && out == 0 {
+				// Grow output buffer, only in the case that it was not large enough for QAT to output any data.
+				if len(z.outputBuf) == DefaultMaxBufferLength {
+					return 0, ErrMaxBufferGrowth
+				}
```
Comment (markbrown314) [inline qatzip/decompress.go:202]: Return produced and bubble up ErrBuffer to caller.

File: qatzip/compress.go [hunk 14]
```diff
@@ -341,19 +341,24 @@ func (z *Writer) compressWrite(p []byte) (n int, err error) {
 
 		if err != nil {
 			if err == ErrBuffer {
-				// expand output buffer
-				// TODO grow to a maximum size
+				// Grow output buffer, only in the case that it was not large enough for QAT to output any data.
+				if z.outputBuf.Len() == DefaultMaxBufferLength {
+					return produced, ErrMaxBufferGrowth
+				}
```
Comment (markbrown314) [inline qatzip/compress.go:347]: Return consumed and bubble up ErrBuffer to caller.

File: qatzip/options.go [hunk 15]
```diff
@@ -183,10 +183,10 @@ func DebugLevelOption(level DebugLevel) Option {
 	}
 }
 
-// Output buffer size (Reader, Writer)
+// Output buffer size (Reader, Writer) (Max buffer size is 2GB)
```
Comment (markbrown314) [inline qatzip/options.go]: Remove: Max buffer size is no longer just 2GB.

File: qatzip/options.go [hunk 16]
```diff
@@ -183,10 +183,10 @@ func DebugLevelOption(level DebugLevel) Option {
...
 func OutputBufLengthOption(len int) Option {
 	return func(a applier) error {
-		if len < MinBufferLength {
+		if len < MinBufferLength || len > DefaultMaxBufferLength {
```
Comment (markbrown314) [inline qatzip/options.go]: Move this into the case statement and check z.p.MaxBufferLength

**Review by markbrown314 — CHANGES_REQUESTED** [review]

File: qatzip/errors.go [hunk 17]
```diff
@@ -45,6 +45,8 @@ var (
 	ErrInputBufferMode         = errors.New(QatErrHdr + "invalid input buffer mode")
 	ErrApplyPostInit           = errors.New(QatErrHdr + "cannot apply options after Reset() or I/O")
 	ErrApplyInvalidType        = errors.New(QatErrHdr + "option appied to incorrect type")
+	ErrMaxBufferGrowth         = errors.New(QatErrHdr + "cannot grow buffer large enough")
```
Comment (markbrown314) [inline qatzip/errors.go:48]: This is not needed anymore. I will remove it in the next release.

**Review by markbrown314 — APPROVED** [review]
Looks good. Approved. I will make a test build for @AlexandruCihodaru to try to see if it resolves issue #1.

**Comment by markbrown314** [comment]
Merging into topic-branch for testing
